tempodb/blocklist/poller.go

```diff
@@ -0,0 +1,162 @@
+	})
+	metricBlocklistLength = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempodb",
+		Name:      "blocklist_length",
+		Help:      "Total number of blocks per tenant.",
+	}, []string{"tenant"})
+)
+
+// PerTenant is a map of tenant ids to backend.BlockMetas
+type PerTenant map[string][]*backend.BlockMeta
+
+// PerTenantCompacted is a map of tenant ids to backend.CompactedBlockMetas
+type PerTenantCompacted map[string][]*backend.CompactedBlockMeta
+
+// Poller retrieves the blocklist
+type Poller struct {
+	reader          backend.Reader
+	compactor       backend.Compactor
+	pollConcurrency uint
+}
+
+// NewPoller creates the Poller
+func NewPoller(pollConcurrency uint, reader backend.Reader, compactor backend.Compactor) *Poller {
+	return &Poller{
+		reader:          reader,
+		compactor:       compactor,
+		pollConcurrency: pollConcurrency,
+	}
+}
+
+// Do does the doing of getting a blocklist
+func (p *Poller) Do() (PerTenant, PerTenantCompacted, error) {
+	start := time.Now()
+	defer func() { metricBlocklistPollDuration.Observe(time.Since(start).Seconds()) }()
+
+	ctx := context.Background()
+	tenants, err := p.reader.Tenants(ctx)
+	if err != nil {
+		metricBlocklistErrors.WithLabelValues("").Inc()
+		return nil, nil, err
+	}
+
+	blocklist := PerTenant{}
+	compactedBlocklist := PerTenantCompacted{}
+
+	for _, tenantID := range tenants {
+		newBlockList, newCompactedBlockList, err := p.pollTenant(ctx, tenantID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		metricBlocklistLength.WithLabelValues(tenantID).Set(float64(len(newBlockList)))
+
+		blocklist[tenantID] = newBlockList
+		compactedBlocklist[tenantID] = newCompactedBlockList
+	}
+
+	return blocklist, compactedBlocklist, nil
+}
+
+func (p *Poller) pollTenant(ctx context.Context, tenantID string) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
+	blockIDs, err := p.reader.Blocks(ctx, tenantID)
+	if err != nil {
+		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
+		return []*backend.BlockMeta{}, []*backend.CompactedBlockMeta{}, err
+	}
+
+	bg := boundedwaitgroup.New(p.pollConcurrency)
+	chMeta := make(chan *backend.BlockMeta, len(blockIDs))
+	chCompactedMeta := make(chan *backend.CompactedBlockMeta, len(blockIDs))
+	anyError := atomic.Error{}
+
+	for _, blockID := range blockIDs {
+		bg.Add(1)
+		go func(uuid uuid.UUID) {
+			defer bg.Done()
+			m, cm, err := p.pollBlock(ctx, tenantID, uuid)
+			if m != nil {
+				chMeta <- m
+			} else if cm != nil {
+				chCompactedMeta <- cm
+			} else if err != nil {
+				anyError.Store(err)
+			}
+		}(blockID)
+	}
+
+	bg.Wait()
+	close(chMeta)
+	close(chCompactedMeta)
+
+	if err = anyError.Load(); err != nil {
+		return nil, nil, err
+	}
+
+	newBlockList := make([]*backend.BlockMeta, 0, len(blockIDs))
+	for m := range chMeta {
+		newBlockList = append(newBlockList, m)
+	}
+	sort.Slice(newBlockList, func(i, j int) bool {
+		return newBlockList[i].StartTime.Before(newBlockList[j].StartTime)
```

Comment:
@joe-elliott Do you recall why we sort by time?  Is this to reduce the work necessary to find recent blocks?  I'm wondering we can sort by ID instead, which might enable faster searching on the blocklist.

Reply:
the only thing i can think of is that in the query frontend this will make it consistently push jobs for certain blocks first which would make the search more likely to return results on one end or another of the range.

the way this is written we are ordering by starttime ascending which makes it most likely to return results at the beginning of the range. if anything i think we'd prefer the opposite and return the most recent results satisfying a particular query.

sorting 10k+ blocks might take a bit and shouldn't be in the query path. i think we should keep this in an order that the query path prefers. however, sorting 10k+ blocks in the polling cycle would be more acceptable.

Reply:
I was thinking of pushing this sort into the blocklist.List struct, perhaps on the ApplyPollingResults ordered by ID.  Per your comment above, are you saying that would impact the query path here negatively?